[JIT] Use Type Level Granularity in Alias Analysis Wildcards

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32326 [JIT] Improve May Contain Alias Using Contained Elements
* **#32251 [JIT] Use Type Level Granularity in Alias Analysis Wildcards**

Previously wildcard sets were associated by TypeKind, meaning all Lists were in one alias set, all Classes were in one alias set, etc. We can improve analysis by bucketing wildcard sets by TypePtr instead. Any two mutable types which can unify should be in the same wildcard set bucket. 

This also allows us do much simpler `mayContainAlias` analysis, and also improves `analyzeConservative` analysis because now we can recurse through all contained memory locations and mark writes, instead of just recursing only level deep in contained elements.

Differential Revision: [D19563263](https://our.internmc.facebook.com/intern/diff/D19563263)

Commit:
Do wildcard sets by typeptr

[ghstack-poisoned]

Commit:
Update on "[JIT] Use Type Level Granularity in Alias Analysis Wildcards"


Previously wildcard sets were associated by TypeKind, meaning all Lists were in one alias set, all Classes were in one alias set, etc. We can improve analysis by bucketing wildcard sets by TypePtr instead. Any two mutable types which can unify should be in the same wildcard set bucket.

Now that we have type-level granularity we can also improve `mayContainAlias` queries. Initialize all new values as containing the wildcard set of each contained mutable type. Whenever a value is added to a container it is set to the wildcard set.

This also improves `analyzeConservative` analysis because now we can recurse through all contained memory locations and mark writes, instead of just recursing only level deep in contained elements.



[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit ac8d14c1:


**None of the build failures appear to be your fault**.

* **1/2** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=8b4c695e4703f39a78028eed5e90051273f74d8d&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 8b4c695e since Jan 15
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **1/2** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=ac8d14c15a49649fe859d2cc902d1ef684c1c0b1).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4263958) pytorch_linux_backward_compatibility_check_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=86735361) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=86735361)) :snowflake:
<details>
<summary>
<code>error pulling image configuration: received unexpected HTTP status: 500 Internal Server Error</code>
</summary>

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3.6-gcc5.4:07597f23-fa81-474c-8bef-5c8a91b50595-ac8d14c15a49649fe859d2cc902d1ef684c1c0b1 
Parallel backend flags:  
error pulling image configuration: received unexpected HTTP status: 500 Internal Server Error 
```

</details>

### 1 failure *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>caffe2_onnx_py3_6_clang7_ubuntu16_04_build</sup> | <sup>Build</sup> | &#128721; [Broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=8b4c695e4703f39a78028eed5e90051273f74d8d&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35&highlight_job=caffe2_onnx_py3_6_clang7_ubuntu16_04_build) |


---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[JIT] Use Type Level Granularity in Alias Analysis Wildcards"


Previously wildcard sets were associated by TypeKind, meaning all Lists were in one alias set, all Classes were in one alias set, etc. We can improve analysis by bucketing wildcard sets by TypePtr instead. Any two mutable types which can unify should be in the same wildcard set bucket.

Now that we have type-level granularity we can also improve `mayContainAlias` queries. Initialize all new values as containing the wildcard set of each contained mutable type. Whenever a value is added to a container it is set to the wildcard set. Then, we can just check if the `containedMemoryLocations` of two sets overlap. 

This also improves `analyzeConservative` analysis because now we can recurse through all contained memory locations and mark writes, instead of just recursing only level deep in contained elements.



[ghstack-poisoned]

Commit:
Update on "[JIT] Use Type Level Granularity in Alias Analysis Wildcards"


Previously wildcard sets were associated by TypeKind, meaning all Lists were in one alias set, all Classes were in one alias set, etc. We can improve analysis by bucketing wildcard sets by TypePtr instead. Any two mutable types which can unify should be in the same wildcard set bucket.

Now that we have type-level granularity we can also improve `mayContainAlias` queries. Initialize all new values as containing the wildcard set of each contained mutable type. Whenever a value is added to a container it is set to the wildcard set. Then, we can just check if the `containedMemoryLocations` of two sets overlap. 

This also improves `analyzeConservative` analysis because now we can recurse through all contained memory locations and mark writes, instead of just recursing only level deep in contained elements.



[ghstack-poisoned]

Commit:
Update on "[JIT] Use Type Level Granularity in Alias Analysis Wildcards"


Previously wildcard sets were associated by TypeKind, meaning all Lists were in one alias set, all Classes were in one alias set, etc. We can improve analysis by bucketing wildcard sets by TypePtr instead. Any two mutable types which can unify should be in the same wildcard set bucket. 

This also allows us do much simpler `mayContainAlias` analysis, and also improves `analyzeConservative` analysis because now we can recurse through all contained memory locations and mark writes, instead of just recursing only level deep in contained elements.

[ghstack-poisoned]

suo commented on torch/csrc/jit/type_hashing.cpp:
```diff
@@ -0,0 +1,28 @@
+namespace torch {
+namespace jit {
+
+size_t HashType::operator()(const TypePtr& type) const noexcept {
```
What's wrong with the default hash?

suo reviewed (approved):
Had some comments inline but nothing blocking

eellison commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -7,30 +7,43 @@
-// introduce more granularity here (e.g. List[int] will never alias
-// List[float]).
-c10::optional<TypeKind> AliasDb::getMutableTypeKind(const TypePtr& type) {
+// For any mutable type, map it to a type such that all other types which it can
```
* Yea, we don't need to unwrap the `Future`, I was following existing codebase.
Both `Future[Tensor]` and `(Tensor, int, Tensor)` are sort of special cases because these are immutable containers. So it doesn't really matter if we don't consider `Optional[Future[Tensor]]` to not alias `Future[Tensor]` so long as the aliasDb correctly considers the contained elements to maybe alias. Same thing with `(Tensor, int, Tensor) aliasing `(Tensor, Tensor, int)`. I'm not sure what the cleanest solution here is.

* Per https://github.com/pytorch/pytorch/pull/32540, Capsule Type doesn't fit in the IR.

* Currently, InterfaceTypes are handled by marking MethodCall's as conservative, so we mark all inputs as written to and add them to the wildcard set, and we cannot reorder nodes around MethodCalls. I think that approach still applies in the current approach, although it could certainly be improved.

I agree that this may require work to maintain. However, I think in our current state we're doing a sort of half-type based alias analysis that poses many of the same problems of this approach (interfaces & classes are currently consider in separate buckets) without getting a lot of the benefits. Despite my recent PR - `unifyTypes` is not a particularly hot piece of code. There aren't that many features we would add in the future that don't pose complications for our existing codebase but do for this PR.

Commit:
Update on "[JIT] Use Type Level Granularity in Alias Analysis Wildcards"


Previously wildcard sets were associated by TypeKind, meaning all Lists were in one alias set, all Classes were in one alias set, etc. We can improve analysis by bucketing wildcard sets by TypePtr instead. Any two mutable types which can unify should be in the same wildcard set bucket. 

This also allows us do much simpler `mayContainAlias` analysis, and also improves `analyzeConservative` analysis because now we can recurse through all contained memory locations and mark writes, instead of just recursing only level deep in contained elements.

[ghstack-poisoned]
